cairo/ethereum/cancun/bloom.cairo

```diff
@@ -0,0 +1,168 @@
+    tempvar hash_bytes = Bytes(new BytesStruct(hash_low_bytes, 16));
+    _add_bloom_index(hash_bytes, 0);
+    _add_bloom_index(hash_bytes, 2);
+    _add_bloom_index(hash_bytes, 4);
+
+    return ();
+}
+
+// @dev does not expect to be called with indexes other than 0, 2, 4
+func _add_bloom_index{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, bloom: MutableBloom}(
+    bloom_entry_hash: Bytes, index: felt
+) {
+    alloc_locals;
+    tempvar hash_subset = Bytes(new BytesStruct(bloom_entry_hash.value.data + index, 2));
+    let hash_subset_uint = Uint_from_be_bytes(hash_subset);
+
+    assert bitwise_ptr.x = hash_subset_uint.value;
+    assert bitwise_ptr.y = BIT_MASK_11_BITS;
+    tempvar bit_to_set = bitwise_ptr.x_and_y;
+    let bitwise_ptr = bitwise_ptr + BitwiseBuiltin.SIZE;
+
+    let bit_index = BIT_MASK_11_BITS - bit_to_set;
+    let (byte_index, bit_position) = divmod(bit_index, 8);
+
+    // Calculate 1 << (7 - bit_position) using multiplication
+    // This is equivalent to 2^bit_shift where bit_shift = 7 - bit_position
+    let (_, bit_shift_value) = divmod(7 - bit_position, 8);
+    let (bit_value) = pow2(bit_shift_value);
+
+    // Read current byte value
+    let bloom_ptr = cast(bloom.value.dict_ptr, DictAccess*);
+    let (current_value) = dict_read{dict_ptr=bloom_ptr}(byte_index);
+
+    // Set the bit using OR operation
+    assert bitwise_ptr.x = current_value;
+    assert bitwise_ptr.y = bit_value;
+    let new_value = bitwise_ptr.x_or_y;
+    let bitwise_ptr = bitwise_ptr + BitwiseBuiltin.SIZE;
+
+    // Write back the updated byte
+    dict_write{dict_ptr=bloom_ptr}(byte_index, new_value);
+
+    tempvar bloom = MutableBloom(
+        new MutableBloomStruct(
+            dict_ptr_start=bloom.value.dict_ptr_start,
+            dict_ptr=cast(bloom_ptr, Bytes1DictAccess*),
+            len=bloom.value.len,
+        ),
+    );
+
+    return ();
+}
+
+func logs_bloom{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr: KeccakBuiltin*}(
+    logs: TupleLog
+) -> Bloom {
+    alloc_locals;
+    let (local mutable_bloom_start) = default_dict_new(0);
+    let mutable_bloom_end = mutable_bloom_start;
+
+    tempvar mutable_bloom = MutableBloom(
+        new MutableBloomStruct(
+            dict_ptr_start=cast(mutable_bloom_start, Bytes1DictAccess*),
+            dict_ptr=cast(mutable_bloom_end, Bytes1DictAccess*),
+            len=256,
+        ),
+    );
+
+    _iter_logs{bloom=mutable_bloom}(logs, 0);
+
+    let (local bloom_buffer: felt*) = alloc();
+    local range_check_ptr = range_check_ptr;
+    local bitwise_ptr: BitwiseBuiltin* = bitwise_ptr;
+    local keccak_ptr: KeccakBuiltin* = keccak_ptr;
+
+    tempvar index = 255;
+    tempvar bloom_end = cast(mutable_bloom.value.dict_ptr, DictAccess*);
+
+    loop:
+    let index = [ap - 2];
+    let dict_ptr = cast([ap - 1], DictAccess*);
+
+    let (value) = dict_read{dict_ptr=dict_ptr}(index);
+    assert bloom_buffer[index] = value;
```

Comment:
why do we need a bloom_buffer?

Reply:
moving values from a mutable DS (dict) to an array